### DISCUSSION TITLE
how to match EOF

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1048] [Q&A] [answered]
Is it possible to match end of file (EOF)? I tried using:

```langium
terminal EOF: /\z/;
```

But it didn't work out for me.

---

Originally, I wanted to create a language in which newlines were important. For example, an assignment statement:

```lang
x=1
y=2
```

If I used:

```langium
hidden terminal WS: /\s+/;
```

It well make the following code valid:

```lang
x=1 y=2
```

So now I've got to add a newline after that assignment statement to make sure that each statement is on a separate line, but if the assignment was at the EOF, it creates an exception asking for a newline ([playground](https://langium.org/playground?grammar=OYJwhgthYgBAMmAdsArmYBTAUNzSAXEAT1gFkB7AE0wBsAubWZ2ANxgEswAjWzAZwDUAXgCC-fh2BII%2BAgCoA3LnGTpswoxZtOPPsICSAEVgByYaZ21UmYQDkAqmQBCAUQBKsABR3XAdXgDX0FYAB9YVwB5ADEASmVsAkwQCA4kMFpYY3pYAHoAbTAAWgAvUSKALQBdQVzlJJS0jNhHFw8c3KKAfnyABiKAThqvLvoAHQA6PsGa2K663AbU9Myo6I6xkoWlpszfAKDXDaQFoA&content=B4XgTAUKAsQ)).

**Comment by msujew** [comment]
I think EOF matching can be accomplished with Langium but is rather complicated. Instead I would recommend the following approach:

```
Model: (variables+=Assignment (NEWLINE+ variables+=Assignment)*)?
```

And remove `NEWLINE+` from the `Assignment` rule.

**Reply by Yokozuna59** [reply]
Thanks!

> I think EOF matching can be accomplished with Langium but is rather complicated.

But is it that complicated? Because that was my original question. If I used your approach for each use case, it would make the grammar longer and more complex. The assignment statement is just an example.

Although I'd go with your answer, I think someone else would ask the same question as me, so I don't think it's appropriate to mark it as answered yet.

**Reply by msujew** [reply]
You could create a dummy token and then return `EOF` from the `chevrotain` import in its stead. You'll need to override the `TokenBuilder` service for that.

**Reply by spoenemann** [reply]
Sounds like we could add `EOF` as a built-in keyword to the grammar language?

**Reply by Yokozuna59** [reply]
> You could create a dummy token and then return `EOF` from the `chevrotain` import in its stead. You'll need to override the `TokenBuilder` service for that.

It looks like it's really complicated, so I need to modify the source code rather than a complex regex in Langium.

> Sounds like we could add `EOF` as a built-in keyword to the grammar language?

I guess it would be great to do so! Since it might be a common use case for languages/files where indentation/newlines are important, e.g., Python, Yaml.

**Reply by Yokozuna59** [reply, marked as answer]
> You could create a dummy token and then return `EOF` from the `chevrotain` import in its stead. You'll need to override the `TokenBuilder` service for that.

Am I missing something?

Added a dummy terminal rule:

```langium
terminal EOF: /\z/;
```

Override `TokenBuilder`:

```typescript
import { TokenType, EOF } from 'chevrotain';
import { DefaultTokenBuilder } from 'langium';
import { TerminalRule } from 'langium/lib/grammar/generated/ast';

class CustomTokenBuilder extends DefaultTokenBuilder {
    override buildTerminalToken(terminal: TerminalRule): TokenType {
        let tokenType = super.buildTerminalToken(terminal);
        if (tokenType.name === 'EOF') {
            tokenType = EOF;
        }
        return tokenType;
    }
}
```

I've also set the `TokenBuilder` as CustomTokenBuilder as shown here:

```typescript
const services = await createServicesForGrammar({
    grammar: grammer,
    module: {
        parser: {
            TokenBuilder: () => new CustomTokenBuilder(),
        },
    },
});
```

**Reply by msujew** [reply]
Are you building your language services purely through the `createServicesForGrammar` function? That one is meant for internal framework tests and more advanced stuff like the Langium playground, but it shouldn't really be used by devs. Do you have a reproducible setup to look into?

Otherwise this seems to look fine. Do you experience any issues with this approach?

**Reply by Yokozuna59** [reply]
> Are you building your language services purely through the `createServicesForGrammar` function? That one is meant for internal framework tests and more advanced stuff like the Langium playground, but it shouldn't really be used by devs.

I'm using it for [unit testing](https://github.com/Yokozuna59/mermaid-lint/blob/2a0ec2bab65554eb9d3f55f452c37b9a1d213461/packages/mermaid-parser/tests/pie/syntax.test.ts); it works just fine, but I'm not sure if it's the best approach.

> Do you have a reproducible setup to look into?
> 
> Otherwise this seems to look fine. Do you experience any issues with this approach?

Here's the actual code [link](https://github.com/Yokozuna59/mermaid-lint/tree/8c27194b46266e19cd8aee4612fda69de380c7a5/packages/mermaid-parser/src/language).

> **Note**: I modified the sixth line [here](https://github.com/Yokozuna59/mermaid-lint/blob/2a0ec2bab65554eb9d3f55f452c37b9a1d213461/packages/mermaid-parser/src/language/grammars/pie.langium#L6) to be:
>
> ```langium
> "pie" (showData?="showData")? (NEWLINE+ | EOF)
> ```

---

I'm currently creating a wrapper grammar that has subgrammars (currently one, though).

If there is a newline, it will show `$type` as the subgrammar (`PieChart`).

```
pie\n
```

But if there's EOF it will show the parent one (`Mermaid`), which isn't as excepted [here](https://github.com/Yokozuna59/mermaid-lint/blob/2a0ec2bab65554eb9d3f55f452c37b9a1d213461/packages/mermaid-parser/tests/pie/syntax.test.ts#L48).

```
pie
```

**Reply by Yokozuna59** [reply]
The thing is that it's working fine for matching stuff, but I keep getting parser and laxer errors, which is why I commented them out. Here is an example.

<https://github.com/Yokozuna59/mermaid-lint/blob/2e5a085cfda71fe60dfa37cc66ff7a1315c459c6/packages/mermaid-parser/tests/common/accessible.test.ts#L285_L286>

**Comment by Yokozuna59** [comment]
After #1162 has been merged, how to actually use it as a rule instead of a value?

I'm trying to create a fragment rule so that could have either `NEWLINE+` or `EOF`, but it keep throwing errors. I tried different ways to implement it but non of them worked.

**Reply by msujew** [reply]
Can you give some more context? Like your grammar and what the error messages are.

**Reply by Yokozuna59** [reply]
```langium
fragment EOL:
  {infer EOL}
  NEWLINE+ | EOF
;
```

It's getting diagnosed in vscode as an error:

![image](https://github.com/eclipse-langium/langium/assets/87622592/147e42b7-c314-4845-9454-bb923aca885e)

I tried generating the AST, it's working as expect.
I tried declaring another rule with the name of `EOF`, but this time AST isn't generated.

**Reply by msujew** [reply]
Uhm, did you reload your VSCode? It likely updated the Langium extension but, wasn't able to reload it yet. You will likely see a `reload required` if you take a look at your installed extensions.

**Reply by Yokozuna59** [reply]
@msujew I did reload VSCode, but I forgot that it was another workspace. Anyway, it isn't diagnosed in VSCode, thanks!

But another issue occurs: if the fragment has been used, the `$type` of the parse result is changed to `EOL`:

![image](https://github.com/eclipse-langium/langium/assets/87622592/0e4f2fc4-d9b7-42a3-ae89-868607c0f1fe)


I guess this is due to `{infer EOL}`. Is there a way to perform the same thing without the need for `{infer EOL}`?

**Reply by msujew** [reply]
I'm not sure why you would want to use `{infer EOL}` at all. It should work as expected without it.

**Reply by Yokozuna59** [reply]
![image](https://github.com/eclipse-langium/langium/assets/87622592/b997f934-b1b3-4318-8da3-a9a77293c583)

```
src/language/common/common.langium:11:10 - This parser rule does not create an object. Add a primitive return type or an action to the start of the rule to force object instantiation.
src/language/common/common.langium:11:1 - Data type rules cannot infer a type.
```

**Reply by msujew** [reply]
Yeah, you should probably make it into a data type rule:

```
EOL returns string: NEWLINE+ | EOF;
```

**Reply by Yokozuna59** [reply]
I tried using the suggested code above, but some test cases were failing. Anyway, I used this approach and it's working fine:

```langium
fragment EOL returns string: NEWLINE+ | EOF;
```

**Reply by Yokozuna59** [reply]
@msujew Sorry to disturb you again, but after removing the overridden lexer, errors were thrown everywhere; it appears that any rule ending with `EOF` would return `undefined` instead of the value of the rule.

This is the grammar: https://github.com/mermaid-js/mermaid/blob/0c574335676b2ba7d5af3d3ac4c6dd1975c1056b/packages/parser/src/language/common/common.langium#L7-L9
And this is code example: https://github.com/mermaid-js/mermaid/blob/0c574335676b2ba7d5af3d3ac4c6dd1975c1056b/packages/parser/tests/pie.test.ts#L37

![image](https://github.com/eclipse-langium/langium/assets/87622592/984c01b1-65c4-4ce0-9ef4-189a890cb448)

This also occurred if the type is array. It would return the variable as `undefined` not each elements in the list.

Should I file an issue?

**Reply by msujew** [reply]
It looks like the EOF token generates an invalid CST node, which then prevents the property assignment from the fragment rule:

![image](https://github.com/eclipse-langium/langium/assets/4377073/e98aed55-434b-4e21-87a3-df8bd8b4c53a)

I'll probably have time to write a fix on Monday and release a bugfix version.

**Reply by msujew** [reply]
I've created an issue for this: https://github.com/eclipse-langium/langium/issues/1262

**Reply by msujew** [reply]
@Yokozuna59 This has been fixed in release 2.1.1 of Langium 👍

**Comment by Yokozuna59** [comment]
Another issue came across: 

![image](https://github.com/eclipse-langium/langium/assets/87622592/4d2123dc-3453-4cbd-b4b7-17b6c13e9548)

The issue only occurred when trying to run the language server; it's generates AST correctly without any issue. Test cases that are run in node (vitest) are passing.

But any grammar that has pattern regex that matches everything break the language server, e.g., here is `JOURNEY_TASK_TITLE` pattern:

```typescript
/([^\n\r:]+?(?=%%)|[^\n\r:]+)/
```

EDIT: the pattern is overridden in `TokenBuilder`.

**Reply by msujew** [reply]
I see, we have a minor issue in our default token builder. See https://github.com/eclipse-langium/langium/pull/1276. You can easily work around it by manually removing the `EOF` token from the array returned by the token builder.

**Reply by msujew** [reply]
A fix for this has been released in `langium@2.1.2`
